I have written these codes as Homework in "Introduction to Bioinformatics" course at the Sharif University of Technology.  
The problem are from [rosalind](https://rosalind.info/problems/). link of each problem has been commented on each file.  
# table of content
1. Alignment:
    - Edit Distance Alignment - [rosalind](https://rosalind.info/problems/edta/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Alignment/edta-Edit%20Distance%20Alignment.py) 
    - Global Alignment with Scoring Matrix and Affine Gap Penalty - [rosalind](https://rosalind.info/problems/gaff/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Alignment/gaff-global%20Alignment%20with%20gap%20penalty.py) 
    - Local Alignment with Affine Gap Penalty - [rosalind](https://rosalind.info/problems/laff/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Alignment/laff-local%20alignment%20with%20gap%20penalty.py) 
    - Finding a Motif with Modifications - [rosalind](https://rosalind.info/problems/sims/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Alignment/sims-Finding%20a%20Motif%20with%20Modifications.py) 
    - Align Two Strings Using Linear Space - [rosalind](https://rosalind.info/problems/ba5l/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Alignment/ba5l-Align%20Two%20Strings%20Using%20Linear%20Space.py) 
 
2. Phylogenetic Tree:
    - Compute Distances Between Leaves - [rosalind](https://rosalind.info/problems/ba7a/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Phylogenetic%20Tree/ba7a-Distance%20between%20Leaves%20Problem.py) 
    - Compute Limb Lengths in a Tree - [rosalind](https://rosalind.info/problems/ba7b/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Phylogenetic%20Tree/ba7b-Compute%20Limb%20Lengths%20in%20a%20Tree.py)  
    - Implement AdditivePhylogeny - [rosalind](https://rosalind.info/problems/ba7c/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Phylogenetic%20Tree/ba7c-Implement%20AdditivePhylogeny.py) 
    - Implement the Neighbor Joining Algorithm - [rosalind](https://rosalind.info/problems/ba7e/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Phylogenetic%20Tree/ba7e-Implement%20the%20Neighbor%20Joining%20Algorithm.py) 
    - Implement UPGMA - [rosalind](https://rosalind.info/problems/ba7d/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Phylogenetic%20Tree/ba7d-Implement%20UPGMA.py) 
 
3. Genome Assembly:
    - Constructing a De Bruijn Graph - [rosalind](https://rosalind.info/problems/dbru/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Genome%20Assembly/dbru-Constructing%20a%20De%20Bruijn%20Graph.py) 
    - Assessing Assembly Quality with N50 and N75 - [rosalind](https://rosalind.info/problems/asmq/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Genome%20Assembly/asmq-Assessing%20Assembly%20Quality%20with%20N50%20and%20N75.py) 
    - Genome Assembly with Perfect Coverage - [rosalind](https://rosalind.info/problems/pcov/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Genome%20Assembly/pcov-Genome%20Assembly%20with%20Perfect%20Coverage.py) 
    - Error Correction in Reads - [rosalind](https://rosalind.info/problems/corr/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Genome%20Assembly/corr-Error%20Correction%20in%20Reads.py) 

4. Motif Finding:
    - Find a Median String - [rosalind](https://rosalind.info/problems/ba2b/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Motif%20Finding/ba2b-Find%20a%20Median%20String.py) 
    - Implement GreedyMotifSearch with Pseudocounts - [rosalind](https://rosalind.info/problems/ba2e/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Motif%20Finding/ba2e-Implement%20GreedyMotifSearch%20with%20Pseudocounts.py) 
    - Implement RandomizedMotifSearch - [rosalind](https://rosalind.info/problems/ba2f/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Motif%20Finding/ba2f-Implement%20RandomizedMotifSearch.py) 
    - Implement GibbsSampler - [rosalind](https://rosalind.info/problems/ba2g/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/Motif%20Finding/ba2g-Implement%20GibbsSampler.py) 

5. HMM:
    - Implement the Viterbi Algorithm - [rosalind](https://rosalind.info/problems/ba10c/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/HMM/ba10c-Implement%20the%20Viterbi%20Algorithm.py) 
    - Implement Baum-Welch Learning - [rosalind](https://rosalind.info/problems/ba10k/) - [my solution](https://github.com/amirrezamirzaei/Bioinformatic_Algorithms_Course/blob/main/HMM/ba10k-Implement%20Baum-Welch%20Learning.py) 
